desktop/ui/src/main/java/org/datacleaner/panels/SchemaTreePanel.java

```diff
@@ -39,73 +48,147 @@
+
+        // load the schema tree in the background because it will retrieve
+        // metadata about the datastore (might take several seconds)
+        new SwingWorker<SchemaTree, Void>() {
+            @Override
+            protected SchemaTree doInBackground() throws Exception {
+                Injector injector = _injectorBuilder.with(Datastore.class, datastore).createInjector();
+                SchemaTree tree = injector.getInstance(SchemaTree.class);
+                return tree;
+            }
+
+            protected void done() {
+                try {
+                    final SchemaTree schemaTree = get();
+                    final JScrollPane schemaTreeScroll = WidgetUtils.scrolleable(schemaTree);
+                    schemaTreeScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
+                    schemaTree.addComponentListener(new ComponentAdapter() {
+                        @Override
+                        public void componentResized(ComponentEvent e) {
+                            updateParentPanel();
+                        }
+
+                    });
+                    removeAll();
+                    add(schemaTreeScroll, BorderLayout.CENTER);
+                    _searchTextField.setText(DEFAULT_SEARCH_FIELD_TEXT);
+                    _searchTextField.addKeyListener(new KeyListener() {
+
+                        @Override
+                        public void keyTyped(KeyEvent e) {
+                        }
+
+                        @Override
+                        public void keyReleased(KeyEvent e) {
+                            JTextField source = (JTextField) e.getSource();
+                            schemaTree.filter(source.getText());
+                        }
+
+                        @Override
+                        public void keyPressed(KeyEvent e) {
+                        }
+                    });
+                    _searchTextField.addFocusListener(new FocusListener() {
+
+                        @Override
+                        public void focusLost(FocusEvent e) {
+                            JTextField source = (JTextField) e.getSource();
+                            if (source.getText().equals("")) {
+                                source.setText(DEFAULT_SEARCH_FIELD_TEXT);
+                            }
+
+                        }
+
+                        @Override
+                        public void focusGained(FocusEvent e) {
+                            JTextField source = (JTextField) e.getSource();
+                            if (source.getText().equals(DEFAULT_SEARCH_FIELD_TEXT)) {
+                                source.setText("");
+                            }
+                        }
+                    });
+                    DCPanel searchPanel = new DCPanel(Color.WHITE);
+                    searchPanel.setLayout(new BorderLayout());
+                    JLabel resetSearchFieldIcon = new JLabel("X");
```

Comment:
Not that it matters much, but I don't consider it to just a label - it's clearly a button IMO. So using a label is a bit strange and also means you have to use a MouseListener instead of a ActionListener (would be less cumbersome I think).